crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs

```diff
@@ -0,0 +1,573 @@
+    query::Access,
+    schedule_v3::{
+        is_apply_system_buffers, BoxedCondition, ExecutorKind, SystemExecutor, SystemSchedule,
+    },
+    system::BoxedSystem,
+    world::World,
+};
+
+/// A funky borrow split of [`SystemSchedule`] required by the [`MultiThreadedExecutor`].
+struct SyncUnsafeSchedule<'a> {
+    systems: &'a [SyncUnsafeCell<BoxedSystem>],
+    conditions: Conditions<'a>,
+}
+
+struct Conditions<'a> {
+    system_conditions: &'a mut [Vec<BoxedCondition>],
+    set_conditions: &'a mut [Vec<BoxedCondition>],
+    sets_of_systems: &'a [FixedBitSet],
+    systems_in_sets: &'a [FixedBitSet],
+}
+
+impl SyncUnsafeSchedule<'_> {
+    fn new(schedule: &mut SystemSchedule) -> SyncUnsafeSchedule<'_> {
+        SyncUnsafeSchedule {
+            systems: SyncUnsafeCell::from_mut(schedule.systems.as_mut_slice()).as_slice_of_cells(),
+            conditions: Conditions {
+                system_conditions: &mut schedule.system_conditions,
+                set_conditions: &mut schedule.set_conditions,
+                sets_of_systems: &schedule.sets_of_systems,
+                systems_in_sets: &schedule.systems_in_sets,
+            },
+        }
+    }
+}
+
+/// Per-system data used by the [`MultiThreadedExecutor`].
+// Copied here because it can't be read from the system when it's running.
+struct SystemTaskMetadata {
+    /// The `ArchetypeComponentId` access of the system.
+    archetype_component_access: Access<ArchetypeComponentId>,
+    /// Indices of the systems that directly depend on the system.
+    dependents: Vec<usize>,
+    /// Is `true` if the system does not access `!Send` data.
+    is_send: bool,
+    /// Is `true` if the system is exclusive.
+    is_exclusive: bool,
+}
+
+/// Runs the schedule using a thread pool. Non-conflicting systems can run in parallel.
+pub struct MultiThreadedExecutor {
+    /// Sends system completion events.
+    sender: Sender<usize>,
+    /// Receives system completion events.
+    receiver: Receiver<usize>,
+    /// Metadata for scheduling and running system tasks.
+    system_task_metadata: Vec<SystemTaskMetadata>,
+    /// Union of the accesses of all currently running systems.
+    active_access: Access<ArchetypeComponentId>,
+    /// Returns `true` if a system with non-`Send` access is running.
+    local_thread_running: bool,
+    /// Returns `true` if an exclusive system is running.
+    exclusive_running: bool,
+    /// The number of systems that are running.
+    num_running_systems: usize,
+    /// The number of systems that have completed.
+    num_completed_systems: usize,
+    /// The number of dependencies each system has that have not completed.
+    num_dependencies_remaining: Vec<usize>,
+    /// System sets whose conditions have been evaluated.
+    evaluated_sets: FixedBitSet,
+    /// Systems that have no remaining dependencies and are waiting to run.
+    ready_systems: FixedBitSet,
+    /// copy of `ready_systems`
+    ready_systems_copy: Option<FixedBitSet>,
```

Comment:
Does this need to be an `Option`? Can we just use an empty FixedBitset instead? (with `std::mem::take`)

Reply:
Oh, I didn't know about `std::mem::take`, I'll make the change.